[1/N] Add BackendOptions class

cccclai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #11865
* #11391
* #11390
* __->__ #11389



Introduce backend option as discussed in https://github.com/pytorch/executorch/discussions/10216

Step 1: Introducd Backend Option class

In later stage, it will be plugged in with the rest of the stack. BackendOptions is pretty much a list of BackendOption, and backend option is a key value pair. The key is a string, and the value can be 3 different types, including bool, string and int.


Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712/)

Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712)

cccclai committed:
[1/N] Add BackendOptions class

Introduce backend option as discussed in https://github.com/pytorch/executorch/discussions/10216

Step 1: Introducd Backend Option class

In later stage, it will be plugged in with the rest of the stack.

Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712/)

[ghstack-poisoned]

cccclai commented:
Previous comment in the old PR: https://github.com/pytorch/executorch/pull/11288

From @JacobSzwejbka : use int64_t for int type.

From @mergennachin: 
> I don't think we need to define a generic BackendOption that takes arbitrary key value. Each backend owners could define their options instead. Vulkan will define its own config structure, Core ML will define its own config structure. Wdyt? 

@JacobSzwejbka, @mcr229, @kimishpatel 

Something like this:

```
BackendInterface { // parent interface
  virtual Error setOption(const BackendOptions& options) = 0;
}

struct BackendOptions {  // Empty parent struct
  virtual ~BackendOptions() = default;
};
```

Each individual backends define their own configs

```
struct VulkanBackendOptions : public BackendOptions {
  const char* metadata[20] = "";
  int thread_count = 1;
  // ... other Vulkan-specific options
};

```

```
VulkanBackend extends BackendInterface {
  Error setOption(const BackendOptions& options) override {
    auto* vulkan_opts = dynamic_cast<const VulkanBackendOptions*>(&options);
  }
}
```

// User code

```
VulkanBackendOptions opts;
opts.thread_count = 5;
auto backend = std::make_unique<VulkanBackend>();
backend->setOption(opts);  // 
```

My response 
> I thought about this idea at the beginning, but it means that users code needs to link to a specific backend, otherwise the runner code will fail to build. Currently the same user code can be backend agnostic, meaning the same code can still run with different backends with/without linking the backend. As the example, the portable runner can work with/without xnnpack https://github.com/pytorch/executorch/blob/main/examples/xnnpack/executor_runner/targets.bzl. 

> It also exposes the backend specific symbol to users, and the contract interface is a bit looser. Previously, The contract is like backend <-> ET <-> backend, and if we change it to this, it can be user <-> backend. And users can pass arbitrary things to backend. Users can run backend->init, backend->execute in their own code too, and it will be out of our control.

> With the proposed solution (https://github.com/pytorch/executorch/discussions/10216), the user code can still be backend agnostic, and the backend option is still optional. We are still on the hook of the contract.

cccclai commented:
From @mergennachin: 
> I don't think we need to define a generic BackendOption that takes arbitrary key value. Each backend owners could define their options instead. Vulkan will define its own config structure, Core ML will define its own config structure. Wdyt? 

@JacobSzwejbka, @mcr229, @kimishpatel 

Something like this:

```
BackendInterface { // parent interface
  virtual Error setOption(const BackendOptions& options) = 0;
}

struct BackendOptions {  // Empty parent struct
  virtual ~BackendOptions() = default;
};
```

Each individual backends define their own configs

```
struct VulkanBackendOptions : public BackendOptions {
  const char* metadata[20] = "";
  int thread_count = 1;
  // ... other Vulkan-specific options
};

```

```
VulkanBackend extends BackendInterface {
  Error setOption(const BackendOptions& options) override {
    auto* vulkan_opts = dynamic_cast<const VulkanBackendOptions*>(&options);
  }
}
```

// User code

```
VulkanBackendOptions opts;
opts.thread_count = 5;
auto backend = std::make_unique<VulkanBackend>();
backend->setOption(opts);  // 
```

cccclai commented:
My response: 

> I thought about this idea at the beginning, but it means that users code needs to link to a specific backend, otherwise the runner code will fail to build. Currently the same user code can be backend agnostic, meaning the same code can still run with different backends with/without linking the backend. As the example, the portable runner can work with/without xnnpack https://github.com/pytorch/executorch/blob/main/examples/xnnpack/executor_runner/targets.bzl. 
> 
> It also exposes the backend specific symbol to users, and the contract interface is a bit looser. Previously, The contract is like backend <-> ET <-> backend, and if we change it to this, it can be user <-> backend. And users can pass arbitrary things to backend. Users can run backend->init, backend->execute in their own code too, and it will be out of our control.
> 
> With the proposed solution (https://github.com/pytorch/executorch/discussions/10216), the user code can still be backend agnostic, and the backend option is still optional. We are still on the hook of the contract.

kimishpatel commented:
I think what @mergennachin suggested makes sense. If user code is setting backend specific option they already have to know what backend they are talking about, unless some options are backend agnostic.

Other piece of the puzzle is how the user will update this. And from #10216 it seems that we will pipe `update` (I would call set_backend_options) method that will set all the options. But part that is not clear to me is how would user indicate which backend they setting this option for. So somehow we need a string identifer for the backends that can be used to figure out which backend's update method be called. @mergennachin in this case we cannot quite have what you proposed. Imagine you have a pointer to an instance of `VulkanBackendOptions`. You pass this to runtime's update method which accepts `BackendOptions*`.  Eventually you want to call VulkanBackend's update options method so you need to cast BackendOptions* to VulkanBackendOptions*. You have two choices either do static_cast or dynamic_cast. The later requires rtti which we dont allow. For static_cast, it has to be safe and you need to know that BackendOption* is actually pointing to VulkanBackendOption. To do that you need to have BackendOptions store tag containing what is the type of the derived class. If we do that, then what you suggest can work.

cccclai committed:
Update on "[1/N] Add BackendOptions class"


Introduce backend option as discussed in https://github.com/pytorch/executorch/discussions/10216

Step 1: Introducd Backend Option class

In later stage, it will be plugged in with the rest of the stack.

Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712/)

[ghstack-poisoned]

cccclai committed:
Update on "[1/N] Add BackendOptions class"


Introduce backend option as discussed in https://github.com/pytorch/executorch/discussions/10216

Step 1: Introducd Backend Option class

In later stage, it will be plugged in with the rest of the stack.

Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712/)

[ghstack-poisoned]

cccclai committed:
Update on "[1/N] Add BackendOptions class"


Introduce backend option as discussed in https://github.com/pytorch/executorch/discussions/10216

Step 1: Introducd Backend Option class

In later stage, it will be plugged in with the rest of the stack.

Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712/)

[ghstack-poisoned]

kimishpatel commented on runtime/backend/backend_options.h:
```diff
@@ -0,0 +1,93 @@
+namespace runtime {
+
+// Strongly-typed option key template
+template <typename T>
```
this class doesnt need to be templated. No use of T

kimishpatel commented on runtime/backend/backend_options.h:
```diff
@@ -0,0 +1,93 @@
+struct BackendOption {
+  const char* key; // key is the name of the backend option, like num_threads,
+                   // enable_profiling, etc
+  OptionValue
+      value; // value is the value of the backend option, like 4, true, etc
```
format?

cccclai replied:
updated

kimishpatel commented on runtime/backend/backend_options.h:
```diff
@@ -0,0 +1,93 @@
+
+  // Type-safe setters
+  template <typename T>
+  void set_option(OptionKey<T> key, T value) {
```
Bassed on suggestion above 
```suggestion
  void set_option(const char [kMaxKeyLength]& key, T value) {
```

kimishpatel commented on runtime/backend/backend_options.h:
```diff
@@ -0,0 +1,93 @@
+  size_t size_; // Current number of options
+};
+
+// Helper functions for creating typed option keys (unchanged)
+constexpr OptionKey<bool> BoolKey(const char* k) {
+  return OptionKey<bool>(k);
+}
+
+constexpr OptionKey<int> IntKey(const char* k) {
+  return OptionKey<int>(k);
+}
+
+constexpr OptionKey<const char*> StrKey(const char* k) {
+  return OptionKey<const char*>(k);
+}
```
Dont think you need these

cccclai replied:
It's just helper functions to simplify users' code.

kimishpatel reviewed (changes requested):
made some suggestions

cccclai commented on runtime/backend/backend_options.h:
```diff
@@ -0,0 +1,93 @@
+                   // enable_profiling, etc
+  OptionValue
+      value; // value is the value of the backend option, like 4, true, etc
+};
```
Users don't need get backend option, but the dispatcher needs it and dispatch to the actual backend.

cccclai replied:
Actually you're right. The backend options map is for dispatching. and each backend will get an array ref of backend options.

cccclai commented on runtime/backend/backend_options.h:
```diff
@@ -0,0 +1,93 @@
+      }
+    }
+    // Add new option if space available
+    if (size_ < MaxCapacity) {
+      options_[size_++] = BackendOption{k, value};
+    }
```
Yeah good point. Will update

cccclai committed:
Update on "[1/N] Add BackendOptions class"




Introduce backend option as discussed in https://github.com/pytorch/executorch/discussions/10216

Step 1: Introducd Backend Option class

In later stage, it will be plugged in with the rest of the stack. BackendOptions is pretty much a list of BackendOption, and backend option is a key value pair. The key is a string, and the value can be 3 different types, including bool, string and int.


Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712/)

Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712)

[ghstack-poisoned]

cccclai committed:
Update on "[1/N] Add BackendOptions class"




Introduce backend option as discussed in https://github.com/pytorch/executorch/discussions/10216

Step 1: Introducd Backend Option class

In later stage, it will be plugged in with the rest of the stack. BackendOptions is pretty much a list of BackendOption, and backend option is a key value pair. The key is a string, and the value can be 3 different types, including bool, string and int.


Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712/)

Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712)

[ghstack-poisoned]

cccclai committed:
Update on "[1/N] Add BackendOptions class"




Introduce backend option as discussed in https://github.com/pytorch/executorch/discussions/10216

Step 1: Introducd Backend Option class

In later stage, it will be plugged in with the rest of the stack. BackendOptions is pretty much a list of BackendOption, and backend option is a key value pair. The key is a string, and the value can be 3 different types, including bool, string and int.


Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712/)

Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712)

[ghstack-poisoned]

mergennachin commented on runtime/backend/options.h:
```diff
@@ -0,0 +1,103 @@
+
+// Union replaced with std::variant
+using OptionValue = std::variant<bool, int, const char*>;
+
```
Document the safety and lifecycle of string types

All string keys and values must have static storage duration (string literals, static const char arrays, or global constants). The BackendOptions class does NOT take ownership of strings.

kimishpatel replied:
For the actual value itself, I think we should do this where std::variant itself holds array of chars rather than pointer to char. Like this https://godbolt.org/z/odTc9ex96

mergennachin commented on runtime/backend/options.h:
```diff
@@ -0,0 +1,103 @@
+  return OptionKey<int>(k);
+}
+
+constexpr OptionKey<const char*> StrKey(const char* k) {
+  return OptionKey<const char*>(k);
+}
```
Can you force to accept only string literals with this?

```
template <size_t N>
constexpr OptionKey<const char*> StrKey(const char (&literal)[N]) {
  return OptionKey<const char*>(literal);
}
```

mergennachin commented on runtime/backend/options.h:
```diff
@@ -0,0 +1,103 @@
+
+  // Type-safe setters
+  template <typename T>
+  void set_option(OptionKey<T> key, T value) {
```
rename this to set_option_impl and make it a private function. And add type safe public APIs to enforce key is always safe string... And value is also safe in the case of string.

```
  // String options - only accepts string literals/arrays
  template <size_t KeyN, size_t ValueN>
  Error set_option(
    const char (&key)[KeyN], 
    const char (&value)[ValueN]
  ) noexcept {
    return set_option_impl(key, static_cast<const char*>(value));
  }
  
  // Non-string options
  template <size_t N>
  Error set_option(const char (&key)[N], bool value) noexcept {
    return set_option_impl(key, value);
  }
  
  template <size_t N>
  Error set_option(const char (&key)[N], int value) noexcept {
    return set_option_impl(key, value);
  }
```

mergennachin reviewed (commented):
See inline

kimishpatel commented on runtime/backend/options.h:
```diff
@@ -0,0 +1,103 @@
+
+// Strongly-typed option key template
+template <typename T>
+struct OptionKey {
```
I dont think we need OptionKey as a separate class. Just use `const char *` in the `BackendOption` class.

However, the thing that worries me the most is by owning `const char*` we are giving users a footgun. They can allocate pointer to the backend option name on stack, pass the pointer around and it will go out-of-scope. If this results in segfault that might be your best option. Worst can lead to undefined behaviors.

I would suggest that we just store `const char[kMaxLegthOfBackendOptionName]`.

kimishpatel replied:
You resolved this comment but I dont see it being addressed

cccclai committed:
Update on "[1/N] Add BackendOptions class"




Introduce backend option as discussed in https://github.com/pytorch/executorch/discussions/10216

Step 1: Introducd Backend Option class

In later stage, it will be plugged in with the rest of the stack. BackendOptions is pretty much a list of BackendOption, and backend option is a key value pair. The key is a string, and the value can be 3 different types, including bool, string and int.


Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712/)

Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712)

[ghstack-poisoned]

cccclai committed:
Update on "[1/N] Add BackendOptions class"




Introduce backend option as discussed in https://github.com/pytorch/executorch/discussions/10216

Step 1: Introducd Backend Option class

In later stage, it will be plugged in with the rest of the stack. BackendOptions is pretty much a list of BackendOption, and backend option is a key value pair. The key is a string, and the value can be 3 different types, including bool, string and int.


Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712/)

Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712)

[ghstack-poisoned]

mergennachin commented on runtime/backend/options.h:
```diff
@@ -0,0 +1,188 @@
+   */
+  template <size_t N>
+  Error set_option(const char (&key)[N], bool value) noexcept {
+    ET_CHECK_MSG(N <= kMaxOptionKeyLength, "Option key is too long");
```
static_assert(N <= kMaxOptionKeyLength, "Option key is too long");

for compile time check, as opposed to runtime check

mergennachin commented on runtime/backend/options.h:
```diff
@@ -0,0 +1,188 @@
+   */
+  template <size_t N>
+  Error set_option(const char (&key)[N], int value) noexcept {
+    ET_CHECK_MSG(N <= kMaxOptionKeyLength, "Option key is too long");
```
.

mergennachin commented on runtime/backend/options.h:
```diff
@@ -0,0 +1,188 @@
+   */
+  template <size_t N>
+  Error set_option(const char (&key)[N], const char* value) noexcept {
+    ET_CHECK_MSG(N <= kMaxOptionKeyLength, "Option key is too long");
```
.

mergennachin commented on runtime/backend/options.h:
```diff
@@ -0,0 +1,188 @@
+    // Add new option if space available
+    if (size_ < MaxCapacity) {
+      BackendOption new_option;
+      strncpy(new_option.key, key, kMaxOptionKeyLength - 1);
+      new_option.key[kMaxOptionKeyLength - 1] = '\0';
```
```
const size_t key_len = std::strlen(key);
const size_t copy_len = std::min(key_len, kMaxOptionKeyLength - 1);
std::memcpy(new_option.key, key, copy_len);
new_option.key[copy_len] = '\0';
```

mergennachin commented on runtime/backend/options.h:
```diff
@@ -0,0 +1,188 @@
+   *
+   * @return A const Span containing all BackendOption entries
+   */
+  executorch::runtime::Span<BackendOption> view() const {
```
```
executorch::runtime::Span<const BackendOption> view() const
```

mergennachin commented on runtime/backend/options.h:
```diff
@@ -0,0 +1,188 @@
+   * Default constructor - initializes with zero options.
+   */
+  BackendOptions() : size_(0) {}
+
```
Potentially have this

```
  /**
   * Copy constructor
   */
  BackendOptions(const BackendOptions& other) : size_(other.size_) {
    for (size_t i = 0; i < size_; ++i) {
      options_[i] = other.options_[i];
    }
  }

  /**
   * Copy assignment operator
   */
  BackendOptions& operator=(const BackendOptions& other) {
    if (this != &other) {
      size_ = other.size_;
      for (size_t i = 0; i < size_; ++i) {
        options_[i] = other.options_[i];
      }
    }
    return *this;
  }
```

mergennachin reviewed (approved):
See inline, approving but inline comments should be addressed

kimishpatel commented:
@mergennachin I am concerned about lifecycle and potential footgun it introduces when setting string values. So I would like to resolve that before we land

mergennachin commented:
@kimishpatel 

Can we just always require string value to be a string literals in the API? 

https://github.com/pytorch/executorch/pull/11389#discussion_r2159218778

kimishpatel commented on runtime/backend/options.h:
```diff
@@ -0,0 +1,188 @@
+   *
+   * @return A mutable Span containing all BackendOption entries
+   */
+  executorch::runtime::Span<BackendOption> view() {
```
maybe make it explicit that this is a mutable view.

kimishpatel commented on runtime/backend/options.h:
```diff
@@ -0,0 +1,188 @@
+   * exist, Error::InvalidArgument if type doesn't match
+   */
+  template <typename T, size_t KeyLen>
+  Error get_option(const char (&key)[KeyLen], T& out) const {
```
same as before on keylen

kimishpatel commented on runtime/backend/test/backend_options_test.cpp:
```diff
@@ -0,0 +1,95 @@
+  // Integer get should succeed with correct value
+  EXPECT_EQ(options.get_option("flag", ival), Error::Ok);
+  EXPECT_EQ(ival, 123);
+}
```
Add tests for mutability of the options now that you have mutable view

cccclai replied:
Added

kimishpatel reviewed (changes requested):
I think we should make option itself have char array instead of char*

kimishpatel commented:
> @kimishpatel
> 
> Can we just always require string value to be a string literals, static const char arrays, or global constants in the API like this?
> 
> [#11389 (comment)](https://github.com/pytorch/executorch/pull/11389#discussion_r2159218778)

Yeah those are fine options. I just want API to make it clear, like what you suggested in the comment (although I think length specifier, for option and value, being template param seems a bit weird. Just have a max longest length as some const)

cccclai committed:
Update on "[1/N] Add BackendOptions class"




Introduce backend option as discussed in https://github.com/pytorch/executorch/discussions/10216

Step 1: Introducd Backend Option class

In later stage, it will be plugged in with the rest of the stack. BackendOptions is pretty much a list of BackendOption, and backend option is a key value pair. The key is a string, and the value can be 3 different types, including bool, string and int.


Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712/)

Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712)

[ghstack-poisoned]

cccclai commented:
New comments are addressed, let me know your thoughts

kimishpatel commented on runtime/backend/options.h:
```diff
@@ -0,0 +1,217 @@
+   * @return Error::Ok on success, Error::InvalidArgument if storage is full
+   */
+  template <size_t N>
+  Error set_option(const char (&key)[N], bool value) noexcept {
```
Do we need N? It will just create N different methods for many possible lengths. My suggestion is to remove N as template param and expect callsites to be something like
```
char my_option_name[kMaxOptionKeyLength] = "use_gpu";
...set_option(my_option_name, True)
```

kimishpatel reviewed (approved):
Looks good. I still think we should not need N or KeyLen template params. So I would suggest to fix that.

cccclai commented:
> > @kimishpatel
> > Can we just always require string value to be a string literals, static const char arrays, or global constants in the API like this?
> > [#11389 (comment)](https://github.com/pytorch/executorch/pull/11389#discussion_r2159218778)
> 
> Yeah those are fine options. I just want API to make it clear, like what you suggested in the comment (although I think length specifier, for option and value, being template param seems a bit weird. Just have a max longest length as some const)

I remove the template N, given that it might generate too many code and causing too much size regression. However the cost is that the key length check is runtime instead of compile time, which I think maybe fine

cccclai committed:
Update on "[1/N] Add BackendOptions class"




Introduce backend option as discussed in https://github.com/pytorch/executorch/discussions/10216

Step 1: Introducd Backend Option class

In later stage, it will be plugged in with the rest of the stack. BackendOptions is pretty much a list of BackendOption, and backend option is a key value pair. The key is a string, and the value can be 3 different types, including bool, string and int.


Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712/)

Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712)

[ghstack-poisoned]

kimishpatel commented:
> > > @kimishpatel
> > > Can we just always require string value to be a string literals, static const char arrays, or global constants in the API like this?
> > > [#11389 (comment)](https://github.com/pytorch/executorch/pull/11389#discussion_r2159218778)
> > 
> > 
> > Yeah those are fine options. I just want API to make it clear, like what you suggested in the comment (although I think length specifier, for option and value, being template param seems a bit weird. Just have a max longest length as some const)
> 
> I remove the template N, given that it might generate too many code and causing too much size regression. However the cost is that the key length check is runtime instead of compile time, which I think maybe fine

There shouldnt be any key lengths check really though? Your fn would be `get_option(const char optin_name[kMaxLength]...)`

kimishpatel commented on runtime/backend/options.h:
```diff
@@ -0,0 +1,220 @@
+   * @return Error::Ok on success, Error::InvalidArgument if storage is full
+   */
+  Error set_option(const char* key, bool value) noexcept {
+    ET_CHECK_MSG(
```
See my comment. I didt mean this

cccclai committed:
Update on "[1/N] Add BackendOptions class"




Introduce backend option as discussed in https://github.com/pytorch/executorch/discussions/10216

Step 1: Introducd Backend Option class

In later stage, it will be plugged in with the rest of the stack. BackendOptions is pretty much a list of BackendOption, and backend option is a key value pair. The key is a string, and the value can be 3 different types, including bool, string and int.


Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712/)

Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712)

[ghstack-poisoned]

cccclai committed:
Update on "[1/N] Add BackendOptions class"




Introduce backend option as discussed in https://github.com/pytorch/executorch/discussions/10216

Step 1: Introducd Backend Option class

In later stage, it will be plugged in with the rest of the stack. BackendOptions is pretty much a list of BackendOption, and backend option is a key value pair. The key is a string, and the value can be 3 different types, including bool, string and int.


Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712/)

Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712)

[ghstack-poisoned]

cccclai committed:
Update on "[1/N] Add BackendOptions class"




Introduce backend option as discussed in https://github.com/pytorch/executorch/discussions/10216

Step 1: Introducd Backend Option class

In later stage, it will be plugged in with the rest of the stack. BackendOptions is pretty much a list of BackendOption, and backend option is a key value pair. The key is a string, and the value can be 3 different types, including bool, string and int.


Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712/)

Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712)

[ghstack-poisoned]

cccclai committed:
Update on "[1/N] Add BackendOptions class"




Introduce backend option as discussed in https://github.com/pytorch/executorch/discussions/10216

Step 1: Introducd Backend Option class

In later stage, it will be plugged in with the rest of the stack. BackendOptions is pretty much a list of BackendOption, and backend option is a key value pair. The key is a string, and the value can be 3 different types, including bool, string and int.


Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712/)

Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712)

[ghstack-poisoned]

cccclai committed:
Update on "[1/N] Add BackendOptions class"




Introduce backend option as discussed in https://github.com/pytorch/executorch/discussions/10216

Step 1: Introducd Backend Option class

In later stage, it will be plugged in with the rest of the stack. BackendOptions is pretty much a list of BackendOption, and backend option is a key value pair. The key is a string, and the value can be 3 different types, including bool, string and int.


Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712/)

Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712)

[ghstack-poisoned]

cccclai committed:
Update on "[1/N] Add BackendOptions class"




Introduce backend option as discussed in https://github.com/pytorch/executorch/discussions/10216

Step 1: Introducd Backend Option class

In later stage, it will be plugged in with the rest of the stack. BackendOptions is pretty much a list of BackendOption, and backend option is a key value pair. The key is a string, and the value can be 3 different types, including bool, string and int.


Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712/)

Differential Revision: [D75993712](https://our.internmc.facebook.com/intern/diff/D75993712)

[ghstack-poisoned]
